### DISCUSSION TITLE
Attempt to be Forward-looking on a New Project

### DISCUSSION THREAD

**Discussion by dylink006** [community/community#126225] [Programming Help]
I'm new to LLMs, but I'm planning to build an application that answers technical questions about my API using a RAG system based on my tech docs (e.g., "how can I configure the API request to wait for the response and to retry if the request times out?"). To summarize ...Goal: answer technical questions (found in my /docs/ subdirectory) so the user doesn't have to search and dig for it. My Plan:

Setup LangChain (are there any tips or tricks I should keep in mind?)
Connect to Weaviate (if there's a better VectorDB to use, I'd love recommendations & rationale)
Connect to ChatGPT 3.5 (pls let me know if there's a better model to use)
Vectorize my /docs/ directory (I've never done this before, but it looks like I need a chunking strategy & embedding model)
Create & style a simple modal UI for the chatbot in Airtable (again, if there's a better/quicker way to do this, I'm all ears)
I also want to make it not simply a Q&A bot so that if the answer returned is not valid to the user's use case, there is a feedback query used to improve the process, giving the tool either more context or showing where the documentation should be expanded/refined.

Thanks in advance for any guidance and/or pitfalls to avoid :-)

**Comment by CopperEagle** [comment]
Hey @dylink006 

Sounds like a solid plan. I recommend first experimenting with each piece individually. Maybe start with a simpler project. For langchain: Have a look at [examples](https://github.com/alphasecio/langchain-examples).  Also [here](https://quickaitutorial.com/langchain-rag-fusion-gpt-4o-python-project-easy-ai-chat-for-your-docs/) is a tutorial that may be relevant for you. Then, what model you use depends on what your usecase is.

- Gemeni 1.5 Pro: Long context (1M+). When processing long video, books etc. this is a good model. Note you pay for usage.
- ChatGPT 3.5 is older but cheap and text only. If thats enough you are good. ChatGPT 4V is multimodal, smarter, more expensive. ChatGPT 4o is even better and lower latency. Note you pay for usage. 128k context. 
- Claude Opus, etc: Smarter than ChatGPT 4V. Again you pay for usage. 200k context.
- Cohere Command R+. Good for RAG and development usage is free. It is fairly good at language. Text only. 128k context.

You can try different ones. Again you should probably look at [this tutorial](https://quickaitutorial.com/langchain-rag-fusion-gpt-4o-python-project-easy-ai-chat-for-your-docs/).

Also, just to repeat. If this is your first venture into LLMs, *start with a simpler project*. A basic one is to be able to chat with one of these models on the command line. Then build up complexity.

**Comment by goldieowner17** [comment]
Hey there! 👋

Thanks for posting in the GitHub Community, @dylink006! We're happy you're here. You are more likely to get a useful response if you are posting your question in the applicable category. In this case, it's Programming Help

Also, you are more likely to get a useful response if you are posting your question(s) in the applicable category and are explicit about what your project entails--giving a few more details might help someone give you a nudge in the right direction.

I've gone ahead and moved this to the correct category for you. Good luck!"
